src/Lean/Server/FileWorker/RequestHandling.lean

```diff
@@ -272,20 +295,38 @@ partial def handleDocumentSymbol (_ : DocumentSymbolParams)
   let t := doc.cmdSnaps.waitAll
   mapTask t fun (snaps, _) => do
     let mut stxs := snaps.map (·.stx)
-    let (syms, _) := toDocumentSymbols doc.meta.text stxs
-    return { syms := syms.toArray }
+    return { syms := toDocumentSymbols doc.meta.text stxs #[] [] }
 where
-  toDocumentSymbols (text : FileMap) (stxs : List Syntax) : List DocumentSymbol × List Syntax :=
+  toDocumentSymbols (text : FileMap) (stxs : List Syntax)
+      (syms : Array DocumentSymbol) (stack : List NamespaceEntry) :
+      Array DocumentSymbol :=
     match stxs with
-    | [] => ([], [])
+    | [] => stack.foldl (fun syms entry => entry.finish text syms none) syms
     | stx::stxs => match stx with
-      | `(namespace $id)  => sectionLikeToDocumentSymbols text stx stxs (id.getId.toString) SymbolKind.namespace id
-      | `(section $(id)?) => sectionLikeToDocumentSymbols text stx stxs ((·.getId.toString) <$> id |>.getD "<section>") SymbolKind.namespace (id.map (·.raw) |>.getD stx)
-      | `(end $(_id)?) => ([], stx::stxs)
+      | `(namespace $id)  =>
+        let entry := { name := id.getId.components', stx, selection := id, prevSiblings := syms }
```

Comment:
OMG, we should definitely rename `components'` to `componentsRev`. Your `foldr` was a complete mystery to me before I realized that, there should be a comment on *why* we store `name` like that.